File: labo/networks/add-protocol/addclient.py
"""ソケットを使って ADD プロトコルを実装したクライアントのスクリプト"""

import socket
import struct


def send_msg(sock, msg):
    """ソケットに指定したバイト列を書き込む関数"""
    # これまでに送信できたバイト数
    total_sent_len = 0
    # 送信したいバイト数
    total_msg_len = len(msg)
    # まだ送信したいデータが残っているか判定する
    while total_sent_len < total_msg_len:
        # ソケットにバイト列を書き込んで、書き込めたバイト数を得る
        sent_len = sock.send(msg[total_sent_len:])
        # まったく書き込めなかったらソケットの接続が終了している
        if sent_len == 0:
            raise RuntimeError('socket conection broken')
        # 書き込めた分を加算する
        total_sent_len += sent_len


def recv_msg(sock, total_msg_size):
    """ソケットから特定のバイト数を読み込む関数"""
    # これまでに受信できたバイト数
    total_recv_size = 0
    # 指定したバイト数を受信できたか判定する
    while total_recv_size < total_msg_size:
        # 残りのバイト列を受信する
        received_chunk = sock.recv(total_msg_size - total_recv_size)
        # 1バイトも読めなかったときはソケットの接続が終了している
        if len(received_chunk) == 0:
            raise RuntimeError('socket connection broken')
        # 受信したバイト数を返す
        yield received_chunk
        # 受信できたバイト数を加算する
        total_recv_size += len(received_chunk)

File: labo/networks/add-protocol/add_client.rs
use std::convert::TryInto;
use std::io::{Read, Write};
use std::net::{Shutdown, TcpStream};

fn recv_msg(mut sock: &TcpStream, total_msg_size: usize) -> Vec<u8> {
    // これまでに受信できたバイト数
    let mut total_recv_size = 0;

    let result = loop {
        let buf_size = total_msg_size - total_recv_size;
        let mut buf = vec![0; buf_size];
        //let received_chunk = sock.take(buf_size as u64).read_to_end(&mut buf);
        match sock.read_exact(&mut buf) {
            Ok(_) => (),
            Err(e) => panic!("socket read failed: {:?}", e),
        };

        println!("buf: {:?}", buf);

        if total_recv_size < total_msg_size {
            break buf;
        }

        if buf.len() == 0 {
            // 1バイトも読めなかったときはソケットの接続が終了している
            panic!("socket connection broken");
        }

        total_recv_size += buf.len();
    };

    result
}

fn send_msg(mut sock: &TcpStream, msg: &Vec<u8>) {
    let mut total_sent_len = 0;

    let total_msg_len = msg.len();

    while total_sent_len < total_msg_len {
        let sent_len = match sock.write(&msg[total_sent_len..]) {
            Ok(v) => v,
            Err(e) => panic!("sent_len write failed!!: {:?}", e),
        };

        if sent_len == 0 {
            panic!("socket connection broken");
        }

        total_sent_len += sent_len;
    }
}

fn main() {
    let stream = match TcpStream::connect("127.0.0.1:54321") {
        Ok(stream) => stream,
        Err(e) => panic!("connection failed!! {:?}", e),
    };

    let operand1: i32 = 1000;
    let operand2: i32 = 2000;
    println!("operand1: {}, operand2: {}", operand1, operand2);
    let mut request_msg = operand1.to_be_bytes().to_vec();
    request_msg.append(&mut operand2.to_be_bytes().to_vec());
    send_msg(&stream, &request_msg);
    println!("sent {:?}", request_msg);
    let received_msg = recv_msg(&stream, 8);
    println!("{:?}", received_msg);

    let result = i64::from_be_bytes((&received_msg[..]).try_into().unwrap());
    println!("{:?}", result);

    match stream.shutdown(Shutdown::Both) {
        Ok(_) => (),
        Err(e) => panic!("shutdown failed!! {:?}", e),
    };
}
